src/common/snapshots.rs

```diff
@@ -61,69 +69,101 @@ pub async fn recover_shard_snapshot(
     snapshot_location: ShardSnapshotLocation,
     snapshot_priority: SnapshotPriority,
 ) -> Result<(), StorageError> {
-    let collection = toc.get_collection(&collection_name).await?;
-    collection.assert_shard_exists(shard_id).await?;
-
-    let download_dir = toc.snapshots_download_tempdir()?;
+    // This future is cancel-safe
+    //
+    // - `download_dir` is handled by `tempfile` and would be deleted on drop
+    // - remote snapshot is downloaded into and would be deleted with the `download_dir`
+
+    cancel::future::spawn_cancel_on_drop(move |cancel| async move {
+        let future = async {
+            let collection = toc.get_collection(&collection_name).await?;
+            collection.assert_shard_exists(shard_id).await?;
+
+            let download_dir = toc.snapshots_download_tempdir()?;
+
+            let (snapshot_path, snapshot_temp_path) = match snapshot_location {
+                ShardSnapshotLocation::Url(url) => {
+                    if !matches!(url.scheme(), "http" | "https") {
+                        let description = format!(
+                            "Invalid snapshot URL {url}: URLs with {} scheme are not supported",
+                            url.scheme(),
+                        );
+
+                        return Err(StorageError::bad_input(description));
+                    }
```

Comment:
I thought that checks like this are handled inside the download function. I don't think we have to do it here as well. Please correct me if I'm wrong.

Reply:
`download_file` allows `file://` URLs. Shard snapshot API does not.

Reply:
Check.

Do you think it would make sense to move all functionality for this into a single module, to prevent spreading it all over the code base?

Maybe we can define some specialized functions like: `download_snapshot`, `file_or_download_snapshot`, ... . Or would that only over complicate things?

I know he had this code before, but I'm only thinking about this now.

Reply:
Sure... but not in this PR :D